File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        vis = [[-1 for _ in range(n)] for _ in range(m)]
        dir = [(0,1),(0,-1),(-1,0),(1,0)]
        def bfs(sx, sy, id):
            queue = deque()
            queue.append((sx,sy,id))
            while len(queue)!=0:
                x, y, z = queue.popleft()
                for (dx, dy) in dir:
                    nx, ny = x + dx, y + dy
                    if nx >= 0 and nx < m and ny >=0 and ny < n and grid[nx][ny]==1 and vis[nx][ny]==-1:
                        vis[nx][ny] = z
                        queue.append((nx, ny, z))
            return
        cnt = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and vis[i][j] == -1:
                    vis[i][j] = cnt
                    bfs(i, j, cnt)
                    cnt += 1
                    
        counts = dict(Counter(item for row in vis for item in row))
        mx = 0
        for i, v in counts.items():
            if i != -1:
                mx = max(mx, v)
        return mx
                    



        

File: Data Structures & Algorithms/rotting-fruit/submission-0.cpp
class Solution {
    void bfs(vector<vector<int>>& grid, int sx, int sy, vector<vector<int>>& vis) {
        int m = grid.size();
        int n = grid[0].size();
        queue<vector<int>> q;
        q.push({sx, sy, 0});
        vector<vector<int>> dir = {{0,1},{1,0},{-1,0},{0,-1}};
        while (!q.empty()) {
            auto v = q.front();
            q.pop();
            for(int i=0;i<4;i++) {
                int nx = v[0] + dir[i][0];
                int ny = v[1] + dir[i][1];
                if (nx >=0 && ny >= 0 && nx < m && ny < n && grid[nx][ny]==1 && (vis[nx][ny]==-1 || vis[nx][ny]>v[2]+1)) {
                    q.push({nx, ny, v[2]+1});
                    vis[nx][ny] = v[2]+1;
                }
            }
        }
        return;
    }

public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> vis(m, vector<int>(n, -1));
        for (int i=0; i<m; i++) {
            for (int j =0; j<n; j++) {
                if (grid[i][j]==2) {
                    bfs(grid, i, j, vis);
                    vis[i][j] = 0;
                }
            }
        }
        int mx = 0;
        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++) {
                if (grid[i][j]==1 and vis[i][j]==-1) return -1;
                mx = max(mx, vis[i][j]);
            }
        }
        return mx;
    }
};
